4-load-balancing-and-auto-scaling-groups.md

```diff
@@ -0,0 +1,389 @@
+
+'Connection Draining' is the time to complete "in-flight" requests while the instance is de-registering or unhealthy.
+
+The ELB stops sending new requests to the instance which is deregistering. 
+
+New connections in to the ELB will be directed to OTHER instances. Existing connections will complete and no new connections will be served to the deregistering instance.
+
+Between 1 - 3600 seconds (1 hour), default is 300 seconds.
+
+Can be disabled by setting the value to 0.
+
+If requests are served quickly, set a low value because your requests are short.
+
+If requests are long, set the value higher so that requests are not interrupted and are able to complete.
+
+## Auto Scaling Group
+
+In real life, the load on a website or application can change. More users = greater load.
+
+Auto-scaling groups allow you to scale out to match increased load, as well as to scale in to match a decreased load.
+
+Ensure we have a minimum or maximum number of machines running - good for budget management.
+
+Automatically Register new instances to a load balancer.
+
+ASGs have the following attributes
+- A launch configuration
+    - AMI + Instance Type
+    - EC2 User Data
+    - EBS Volumes
+    - Security Groups
+    - SSH Key Pair
+- Min Size / Max Size & initial Capacity (as well as desired capacity)
+- Networks and Subnet Information
+- Load Balancer + Target Group Information
+- Scaling Policies (what triggers scale out/in)
+
+### Auto Scaling Alarms
+
+Can scale an ASG based on CloudWatch Alarms
+
+Alarms monitor a metric (such as average CPU)
+
+Metrics are computed for the OVERALL ASG INSTANCES
+   
+It's possible to define better auto-scaling rules that are directly managed by EC2
+- Target Average CPU usage
+- Number of requests on the ELB per instance
+- Average Network In
+- Average Network Out
+
+Can auto scale based on a custom metric, e.g. number of connected users
+1. Send a custom metric from application on EC2 to CloudWatch (PutMetric API)
+2. Create CloudWatch alarm to react to low/high values
+3. Use the CloudWatch alarm as the scaling policy for ASG
+
+#### ASG Exam Tips
+
+Scaling policies can be on CPU, Network and even based on custom metrics or a schedule (if you know visitor patterns)
+
+ASGs use Launch Configurations or Launch Templates (newer version of launch config)
+
+To update an ASG, you must provide a new launch configuration / launch template. Underlying EC2s can be replaced over time.
+
+Attaching an IAM role to an ASG will automatically assign the IAM role to the EC2.
+
+ASGs are free. You pay for the underlying resources that are launched.
+
+For instances in an ASG, if an instance is terminated, the ASG will realise that this instance has been removed and launch a new instance to replace this, maintaining your desired capacity.
+
+ASGs can terminate instances marked as unhealthy by a load balancer.
+
+### ASG Scaling Policies
+
+3 Kinds of policies:
+
+Target Tracking Scaling
+- Simple and easy to set up
```

Comment:
With target tracking scaling policies, you select a scaling metric and set a target value.